feat(autoscaling): support warm pools

jungseoklee opened the pull request:
*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### Summary

This PR adds an interface to create a warm pool easily, which gives an ability to decrease latency for applications that have exceptionally long boot times: https://docs.aws.amazon.com/autoscaling/ec2/userguide/ec2-auto-scaling-warm-pools.html

### Test

- Run unit test and integration test.
```
PASS test/warm-pool.test.js
...
=============================== Coverage summary ===============================
Statements   : 93.22% ( 413/443 )
Branches     : 87% ( 288/331 )
Functions    : 93.39% ( 99/106 )
Lines        : 92.95% ( 396/426 )
================================================================================
...
Verifying integ.warm-pool.js against integ.warm-pool.expected.json ... OK.
...
```

### Notes
- I borrowed description from [AWS public doc](https://docs.aws.amazon.com/autoscaling/ec2/APIReference/API_PutWarmPool.html) for docstring.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 2e4c79ae02094f0feabe64918ee756ae664642d5
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=66826e97-ac93-4f13-9221-2b2cfbfe1605%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 559fcdf0a001707c786a096d24661dcd799ec3ef
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e88cab4c-fad6-4026-a9da-7b3cc6f3ce72%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 35f0360c45b6e5022cac77e98a997c3dfbff7d2f
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=864e8a39-77a7-40a9-b7c3-48e273c3654e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

jungseoklee commented:
Here are changes since the first revision:

- Added document link on hibernation prerequisites.
- Rebased on top of the latest HEAD of master branch.

kaizen3031593 commented on packages/@aws-cdk/aws-autoscaling/lib/warm-pool.ts:
```diff
@@ -0,0 +1,101 @@
+  /**
+   * Stopped
+   */
+  STOPPED = 'Stopped'
```
```suggestion
  STOPPED = 'Stopped',
```

kaizen3031593 commented on packages/@aws-cdk/aws-autoscaling/package.json:
```diff
@@ -122,6 +122,7 @@
       "export:@aws-cdk/aws-autoscaling.IAutoScalingGroup",
       "props-physical-name:@aws-cdk/aws-autoscaling.AutoScalingGroupProps",
       "props-physical-name:@aws-cdk/aws-autoscaling.ScheduledActionProps",
+      "props-physical-name:@aws-cdk/aws-autoscaling.WarmPoolProps",
```
Please write a doc for this instead of excluding.

jungseoklee replied:
Will write a doc.

jungseoklee replied:
I think we need to keep it. The following is an error message when I remove this line:

```
error: [awslint:props-physical-name:@aws-cdk/aws-autoscaling.WarmPoolProps] Every Resource must have a single physical name construction property, with a name that is an ending substring of <cfnResource>Name
```

An user cannot assign a name to an warm pool unlike other resources, i.e. lifecycle hooks.

kaizen3031593 replied:
Good call. Keep it as is.

kaizen3031593 commented on packages/@aws-cdk/aws-autoscaling/test/integ.warm-pool.ts:
```diff
@@ -0,0 +1,21 @@
+import * as ec2 from '@aws-cdk/aws-ec2';
+import * as cdk from '@aws-cdk/core';
+import * as autoscaling from '../lib';
+
```
can you add reproduction steps to show that the integ test not only deploys but can be interfaced via api? take a look at some of our other integ tests, they _should_ have reproduction steps.

jungseoklee replied:
Will do.

jungseoklee replied:
Actually, can you elaborate what " reproduction step" mean here? Can you share an example?

kaizen3031593 replied:
Example: https://github.com/aws/aws-cdk/blob/8dd93a8936ddc74b799e0c6dafbe0511ee687e71/packages/@aws-cdk/aws-synthetics/test/integ.canary.ts#L14

kaizen3031593 commented on packages/@aws-cdk/aws-autoscaling/lib/warm-pool.ts:
```diff
@@ -0,0 +1,101 @@
+/**
+ * Basic properties for a warm pool
+ */
+export interface BasicWarmPoolProps {
```
```suggestion
export interface WarmPoolOptions {
```

jungseoklee replied:
To make sure, the suggestion does not align with what this package has in terms of naming convention. Here is [an example](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-autoscaling/lib/lifecycle-hook.ts). Do we want to diverge?

kaizen3031593 replied:
Yes, lets. Looks like the example you linked was introduced over 3 years ago, which might be before we introduced the `Options` -> `Props` paradigm.

jungseoklee replied:
Thanks for confirmation. Will follow the suggestion.

kaizen3031593 commented on packages/@aws-cdk/aws-autoscaling/lib/warm-pool.ts:
```diff
@@ -0,0 +1,101 @@
+   *
+   * @default - false
+   */
+  readonly reuseOnScaleIn?: boolean
```
```suggestion
  readonly reuseOnScaleIn?: boolean;
```

jungseoklee replied:
Will update in the next revision.

kaizen3031593 commented on packages/@aws-cdk/aws-autoscaling/lib/warm-pool.ts:
```diff
@@ -0,0 +1,101 @@
+        reuseOnScaleIn: props.reuseOnScaleIn,
+      } : undefined,
+      maxGroupPreparedCapacity: props.maxGroupPreparedCapacity,
+      minSize: props.minSize && props.minSize,
+      poolState: props.poolState && props.poolState,
```
Why are these set like this? I would have expected something like

```ts
poolState: props.poolState ?? PoolState.STOPPED,
```

jungseoklee replied:
Not to populate `poolState` field in generated CFN template which lets service decide default behavior.

kaizen3031593 commented on packages/@aws-cdk/aws-autoscaling/lib/warm-pool.ts:
```diff
@@ -0,0 +1,101 @@
+  /**
+   * Hibernated
+   *
+   * To use this state, prerequisites must be in place: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/hibernating-prerequisites.html
```
```suggestion
   * To use this state, prerequisites must be in place.
   * @see https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/hibernating-prerequisites.html
```

jungseoklee replied:
Will update in the next revision.

kaizen3031593 commented on packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:
```diff
@@ -744,6 +745,16 @@ abstract class AutoScalingGroupBase extends Resource implements IAutoScalingGrou
+  /**
+   * Add a pool of pre-initialized EC2 instances that sits alongside an Auto Scaling group
+   */
+  public addWarmPool(id: string, props: BasicWarmPoolProps): WarmPool {
```
```suggestion
  public addWarmPool(options?: WarmPoolOptions): WarmPool {
```

kaizen3031593 replied:
Instead of requiring an `id`, lets create it for the user instead. We do this all the time for the constructid elsewhere in the CDK.

jungseoklee replied:
Sounds good. Will update in the next revision.

kaizen3031593 reviewed (changes requested):
Hi @jungseoklee, this looks like a good start. A few preliminary comments for my first iteration.

kaizen3031593 commented on packages/@aws-cdk/aws-autoscaling/README.md:
````diff
@@ -488,6 +488,16 @@ const aspect = new autoscaling.AutoScalingGroupRequireImdsv2Aspect();
+```ts
+declare const autoScalingGroup: autoscaling.AutoScalingGroup;
+
+autoScalingGroup.addWarmPool('Pool', {});
````
```suggestion
autoScalingGroup.addWarmPool('Pool');
```

Reminder that when we change the `addWarmPool` signature this should change as well. I think it might be beneficial to have a second example of `addWarmPool` to show what optional parameters you can set.

jungseoklee replied:
Sure, will do.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 41ac57ff1686ef37e663aeac60dbe56fbb1c3e53
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=62060c7a-a1c9-49dc-af0e-4c59361ed4e3%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

jungseoklee commented:
The PR Linter failure is related to [this commit](https://github.com/aws/aws-cdk/commit/2d2a34d60df5da50d9782ac4405295f070d7a167). Could CDK maintainer check if the following regex is correct?

```
/^integ.*.ts$/
```

Do we need to change regex as below because of file path?

```
/integ.*.ts$/
```

jungseoklee commented:
Here are changes since the second revision:

- Addressed comments from Kaizen.
- Rebased on top of the latest HEAD of master branch.

kaizen3031593 commented on packages/@aws-cdk/aws-autoscaling/lib/warm-pool.ts:
```diff
@@ -0,0 +1,104 @@
+  /**
+   * The instance state to transition to after the lifecycle actions are complete.
+   *
+   * @default - STOPPED
```
```suggestion
   * @default PoolState.STOPPED
```

kaizen3031593 commented on packages/@aws-cdk/aws-autoscaling/lib/warm-pool.ts:
```diff
@@ -0,0 +1,104 @@
+        reuseOnScaleIn: props.reuseOnScaleIn,
+      } : undefined,
+      maxGroupPreparedCapacity: props.maxGroupPreparedCapacity,
+      minSize: props.minSize && props.minSize,
+      poolState: props.poolState && props.poolState,
```
```suggestion
      minSize: props.minSize,
      poolState: props.poolState,
```

If `props.minSize` is undefined the property will be automatically dropped from the CFN template.

jungseoklee replied:
That's great. I am okay to commit this suggestion.

kaizen3031593 reviewed (changes requested):
@jungseoklee This is almost good to go. I will update the defaults myself, but leave the conversation about `minSize` and `poolState` up for you to see/discuss

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 75b6e1918db4a332696fce95246c0865023c6bf8
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=cc4b6b0d-0987-421f-86bc-bc476aa5da00%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 75b6e1918db4a332696fce95246c0865023c6bf8
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=99f05ec1-aefc-4c70-84e7-6e9848175ea0%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

kaizen3031593 reviewed (approved):
Thanks @jungseoklee!

jungseoklee commented:
> Thanks @jungseoklee!

Thanks for reviewing the PR, @kaizen3031593 !

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 123f782ccdbdd6adf3946068fe368ebaca93f67f
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4959c78b-68ec-4ea5-9683-64d1b120fe71%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 1aa59636c149ef1974d86e499929d3ff43ccb1f8
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=6993036e-e6c2-48e7-ad54-a76cc8e5a696%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: a1b1628fc384ba202bee02c00fb36399e5ce0ac5
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=83203230-a5a1-438d-9523-6eb7ad45535a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
